### ISSUE TITLE
Encode JSON arrays as strings in stage options

### ISSUE THREAD

**Issue by chambbj** [PDAL/PDAL#1434]
@abellgithub We discussed this briefly in person last week, and I don't think it's as much of a breaking change as maybe we'd suspected. Interested to get your thoughts though.

For interested readers, I'm looking for the ability to provide stage options as arbitrary JSON objects, or more specifically, I'm looking to supply an array of objects (though there could be an argument to support the former as well).

An example of this is shown in the pipeline below, where we'd like to supply a list of PCL Block supported methods inline with the PDAL pipeline (in PDAL master, this is supplied as a sidecar JSON file or as stringified JSON).

```
{
    "pipeline": [
        "input.las",
        {
            "type":"filters.pclblock",
            "methods": [
                {
                    "name": "PassThrough",
                    "setFilterFieldName": "z",
                    "setFilterLimits":
                    {
                        "min": 410.0,
                        "max": 440.0
                    }
                }
            ]
        },
        "output.las"
    ]
}
```

The only modification that would be required to PDAL is in PipelineReaderJSON.cpp, in `extractOptions`, to allow

```
if (node[name].isArray())
{
    Json::FastWriter w;
    std::string s(w.write(node[name]));
    options.add(name, s);
}
```

In other words, if the option is an array, write it to a string, and add the string as a stage option. In my case, the PCL Block filter will parse this string as it's own self-contained JSON document and construct the PCL pipeline. (Currently an array would just throw an error.)

So, I've done just enough poking at things locally to know that this works for the PCL Block. Question is whether or not there are any ripple effects I'm overlooking (tests pass anyway).

**Comment by msmitherdc** [comment]
@chambbj would this be a place where includes would make sense, including an external json file so that the same plugin options could be used across multiple pipelines?

**Comment by chambbj** [comment]
So, the PCL Block filter already allows you to specify a file for the PCL portion of the pipeline. That could definitely stick around, but I like the idea of being able to specify it inline as well. Makes it easier to read and keep track of.

**Comment by msmitherdc** [comment]
i meant as a general case for pipelines. @hobu has highlighted the need/want to have non pdal json for plugins etc. It would be good to have them both inline or included from external.

**Comment by hobu** [comment]
#1427 is about providing a mechanism to allow any external application who's using PDAL pipeline to extend it in a way that PDAL won't ever care about. If PDAL needs to care about something in some application's extension, it needs to be brought into the PDAL pipeline specification proper.

I'm 👍  on this particular addition if it allows PCL block filters to coexist without much disruption.

**Comment by abellgithub** [comment]
If you're fine with getting this as a string, that seems fine.  What happens when you dump the pipeline?

**Comment by abellgithub** [comment]
On second thought, I'm not so sure.  It's like we're turning over an entire type to some other application that we hope may use it in a way that makes sense.  IDK.  What if PDAL wants to use an array as something in its pipeline specification for options of its own?

**Comment by abellgithub** [comment]
Maybe better to handle it as a keyword?

**Comment by chambbj** [comment]
A keyword to effectively say give me this option as a string? Seems anything more plugin specific (my PCL example) would be getting into the weeds and not something PDAL would want to manage either.

**Comment by abellgithub** [comment]
Maybe best as you proposed.

**Comment by chambbj** [comment]
We could add this to https://github.com/PDAL/PDAL/blob/master/doc/pipeline.rst.

> A stage object may have a member with the name `as_string` whose value will be stringified.

In my example, the PCLBlock stage would be specified as

```
{
    "type": "filters.pclblock",
    "as_string": [{
        "name": "PassThrough",
        "setFilterFieldName": "z",
        "setFilterLimits": {
            "min": 410.0,
            "max": 440.0
        }
    }]
}
```

A more complex, albeit contrived example

```
{
    "type": "filters.pclblock",
    "as_string": [{
        "name": "PassThrough",
        "setFilterFieldName": "z",
        "setFilterLimits": {
            "min": 410.0,
            "max": 440.0
        }
    }, {
        "name": "StatisticalOutlierRemoval",
        "setMeanK": 8,
        "setStddevMulThresh": 0.2
    }, {
        "name": "VoxelGrid",
        "setLeafSize": {
            "x": 1.0,
            "y": 2.0,
            "z": 3.0
        }
    }, {
        "name": "ProgressiveMorphologicalFilter",
        "setMaxDistance": 200.0,
        "setSlope": 0.7
    }]
}
```

In this manner, PDAL's pipeline reader can simply pass along the embedded JSON for subsequent parsing in the stage, without making blanket assumption based on type. At the same time, it doesn't preclude future usage of things like arrays or objects as stage options. For example,

```
{
    "type": "filters.foo",
    "some_array": [
        "hello",
        "world"
    ]
}
```

or

```
{
    "type": "filters.bar",
    "some_object": {
        "hello": "world"
    }
}
```

**Comment by abellgithub** [comment]
I think I like this.  Let me think about it for a bit to see if something comes to mind that makes this ungood.  Right now all this stuff goes through the options system.  Would you call the option "as_string" and just pass it that way?

**Comment by hobu** [comment]
is there a "standard json" way to say "as_string"? The intent is clear, but the name is a bit ick. 

How is this different than a PDAL-consumable version of `extension` like proposed in #1427? I wonder if we shouldn't square these two behaviors and their definitions...

**Comment by connormanning** [comment]
I give a big +1 to the original proposal.  I think the keyworded `as_string` is a step back.  If we have technical limitations that make it impossible, then maybe this is a good time to address them (maybe after the release).

From the perspective of a user, there's no reason to treat JSON options as "special" in a JSON API, hence my -1 to `as_string`.  It seems incredibly unintuitive, and you end up with a meaningless option name that conveys nothing about the option itself - just a sentinel to PDAL.  Sure you could nest your semantic names within `as_string`.  But that seems just as nonsensical as it would be to group all `integer` options under a common name just because they share the same type.

Personally I'd like to see something like `if (node[name].isArray() || node[name].isObject())` invoking an "under-the-hood" stringification that could be accessed as JSON.

**Comment by hobu** [comment]
Another JSON thing I want to be able to provide as options are GeoJSON objects for geometries.

**Comment by abellgithub** [comment]
At this point I have no opinion, so disregard anything I may have said above.

**Comment by chambbj** [comment]
Closed via #1435
